sdk/src/azure/core/az_json_parser.c

```diff
@@ -1,558 +1,693 @@
-  AZ_RETURN_IF_FAILED(_az_is_expected_span(reader, AZ_SPAN_FROM_STR("\"")));
-  AZ_RETURN_IF_FAILED(az_span_reader_get_json_string_rest(reader, &out_token_member->name));
-  *reader = _az_span_trim_white_space(*reader);
-  AZ_RETURN_IF_FAILED(_az_is_expected_span(reader, AZ_SPAN_FROM_STR(":")));
-  *reader = _az_span_trim_white_space(*reader);
-  AZ_RETURN_IF_FAILED(az_json_parser_get_value_space(json_parser, &out_token_member->token));
-  return az_json_parser_check_item_end(json_parser, out_token_member->token);
-}
+  bool within_object
+      = _az_json_stack_peek(&json_parser->_internal.bit_stack) == _az_JSON_STACK_OBJECT;
 
-AZ_NODISCARD az_result
-az_json_parser_parse_array_item(az_json_parser* json_parser, az_json_token* out_token)
-{
-  _az_PRECONDITION_NOT_NULL(json_parser);
-  _az_PRECONDITION_NOT_NULL(out_token);
+  if (next_byte == ',')
+  {
+    json_parser->_internal.bytes_consumed++;
 
-  AZ_RETURN_IF_FAILED(az_json_parser_check_item_begin(json_parser, _az_JSON_STACK_ARRAY));
-  AZ_RETURN_IF_FAILED(az_json_parser_get_value_space(json_parser, out_token));
-  return az_json_parser_check_item_end(json_parser, *out_token);
-}
+    az_span json = _az_json_parser_skip_white_space(json_parser);
 
-AZ_NODISCARD az_result az_json_parser_done(az_json_parser* json_parser)
-{
-  _az_PRECONDITION_NOT_NULL(json_parser);
+    // Expected start of a property name or value, but instead reached end of data.
+    if (az_span_size(json) < 1)
+    {
+      return AZ_ERROR_EOF;
+    }
+
+    next_byte = az_span_ptr(json)[0];
 
-  if (az_span_size(json_parser->_internal.reader) > 0
-      || !az_json_parser_stack_is_empty(json_parser))
+    if (within_object)
+    {
+      // Expected start of a property name after the comma since we are within a JSON object.
+      if (next_byte != '"')
+      {
+        return AZ_ERROR_PARSER_UNEXPECTED_CHAR;
+      }
+      return _az_json_parser_process_property_name(json_parser);
+    }
+    else
+    {
+      return _az_json_parser_process_value(json_parser, next_byte);
+    }
+  }
+  else if (next_byte == '}')
   {
-    return AZ_ERROR_JSON_INVALID_STATE;
+    return _az_json_parser_process_container_end(json_parser, AZ_JSON_TOKEN_END_OBJECT);
+  }
+  else if (next_byte == ']')
+  {
+    return _az_json_parser_process_container_end(json_parser, AZ_JSON_TOKEN_END_ARRAY);
+  }
+  else
+  {
+    // No other character is a valid token delimiter within JSON.
+    return AZ_ERROR_PARSER_UNEXPECTED_CHAR;
   }
-  return AZ_OK;
 }
 
-AZ_NODISCARD az_result
-az_json_parser_skip_children(az_json_parser* json_parser, az_json_token token)
+AZ_NODISCARD az_result az_json_parser_move_to_next_token(az_json_parser* json_parser)
 {
   _az_PRECONDITION_NOT_NULL(json_parser);
 
-  switch (token.kind)
+  az_span json = _az_json_parser_skip_white_space(json_parser);
+
+  // An empty JSON payload is invalid.
+  // TODO: Should this return invalid state instead?
+  if (az_span_size(json) < 1)
```

Comment:
@JeffreyRichter - what are your thoughts on the error `az_result` we should return in this case?

Reply:
spans with negative sizes are non-valid, they violate the contract, correct? Unless there are exception and we are putting new meanings to it. if so, then the code should act as if the size is never negative, i.e. do not check it in the normal `if`s, but to additionally ensure that and detect if our assumptions are incorrect, we put a precondition that he span is valid.

Reply:
This check isn't about negative size'd spans, but rather empty spans, essentially equivalent to `if (az_span_size(json) == 0)`

Reply:
Oh, sorry, I've read it as < 0